src/util/pg.py

```diff
@@ -112,23 +100,57 @@ def execute(query):
 
         cr.commit()
 
+        CONCURRENCY_ERRORCODES = {
+            errorcodes.DEADLOCK_DETECTED,
+            errorcodes.SERIALIZATION_FAILURE,
+        }
+        failed_queries = []
+        tot_cnt = 0
         with ThreadPoolExecutor(max_workers=max_workers) as executor:
-            return sum(
-                log_progress(
-                    executor.map(execute, queries),
-                    logger,
-                    qualifier="queries",
-                    size=len(queries),
-                    estimate=False,
-                    log_hundred_percent=True,
-                )
-            )
+            future_queries = {executor.submit(execute, q): q for q in queries}
+            for future in log_progress(
+                concurrent.futures.as_completed(future_queries),
+                logger,
+                qualifier="queries",
+                size=len(queries),
+                estimate=False,
+                log_hundred_percent=True,
+            ):
+                try:
+                    tot_cnt += future.result() or 0
+                except psycopg2.OperationalError as exc:
+                    if exc.pgcode not in CONCURRENCY_ERRORCODES:
+                        raise
+
+                    # to be retried without concurrency
+                    failed_queries.append(future_queries[future])
+
+        if failed_queries:
+            logger.info("Serialize queries that failed due to concurrency issues")
+            tot_cnt += _parallel_execute_serial(cr, failed_queries, logger=logger)
+            cr.commit()
+
+        return tot_cnt
 
 else:
     _parallel_execute_threaded = _parallel_execute_serial
 
 
 def parallel_execute(cr, queries, logger=_logger):
+    """
+    Try to execute queries concurrently using `min([8, len(queries), n_cpus])` workers.
+    If concurrency issues are detected, some of the queries will be executed sequentially.
+    Side effect: the given cursor is committed.
+
+    Potential gains example (using 8 workers):
+        +---------------------------------------------+-------------+-------------+
+        | File                                        | Sequential  | Parallel    |
+        +---------------------------------------------+-------------+-------------+
+        | base/saas~12.5.1.3/pre-20-models.py         | ~8 minutes  | ~2 minutes  |
+        | mail/saas~12.5.1.0/pre-migrate.py           | ~10 minutes | ~4 minutes  |
+        | mass_mailing/saas~12.5.2.0/pre-10-models.py | ~40 minutes | ~18 minutes |
+        +---------------------------------------------+-------------+-------------+
+    """
```

Comment:
We don't want this to be on the online docs if the automatic docs generation is finally merged.
```suggestion
    :meta private: exclude from online docs
    """
```

Once done, then we can revisit this docstring and harmonize it with the docs PR. The same for any other utils that do not have yet clear docs, i.e. those not touched by the docs PR.